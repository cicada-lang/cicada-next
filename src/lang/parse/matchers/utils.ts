import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
