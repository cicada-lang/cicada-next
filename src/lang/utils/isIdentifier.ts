const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
