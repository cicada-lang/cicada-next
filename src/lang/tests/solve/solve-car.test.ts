import { expect, test } from "vitest"
import { runCode } from "../utils"

test("solve Car", async () => {
  const output = await runCode(`

solve (ab: Pair(String, String), a: String) {
  unify a = car(ab)
  unify ab = cons("a", "b")
}

`)

  expect(output).toMatchInlineSnapshot('"{ ab: cons(\\"a\\", \\"b\\"), a: car(ab) }"')
})
