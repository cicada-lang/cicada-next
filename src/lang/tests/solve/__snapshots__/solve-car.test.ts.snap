// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`solve Car > 
    "{
      ab: cons(\\"a\\", \\"b\\"),
      a: \\"a\\"
    }"
   1`] = `
"{
  ab: cons(\\"a\\", \\"b\\"),
  a: \\"a\\"
}"
`;

exports[`solve Car -- car cons > 
    "{
      a: ?a
    }"
   1`] = `
"{
  a: ?a
}"
`;

exports[`solve Car -- car cons: 
    "{
      a: ?a
    }"
   1`] = `
"{
  a: ?a
}"
`;

exports[`solve Car -- nested > 
    "{
      ab: cons(\\"a\\", \\"b\\"),
      ba: cons(\\"b\\", \\"a\\"),
      a: \\"a\\",
      b: \\"b\\",
      b2: ?b
    }"
   1`] = `
"{
  ab: cons(\\"a\\", \\"b\\"),
  ba: cons(\\"b\\", \\"a\\"),
  a: \\"a\\",
  b: \\"b\\",
  b2: ?b
}"
`;

exports[`solve Car -- nested: 
    "{
      ab: cons("a", "b"),
      ba: cons("b", "a"),
      a: "a",
      b: "b",
      b2: ?b
    }"
   1`] = `
"{
  ab: cons("a", "b"),
  ba: cons("b", "a"),
  a: "a",
  b: "b",
  b2: ?b
}"
`;

exports[`solve Car: 
    "{
      ab: cons("a", "b"),
      a: "a"
    }"
   1`] = `
"{
  ab: cons("a", "b"),
  a: "a"
}"
`;

exports[`solve Car 1`] = `
"{
  ab: cons("a", "b"),
  a: "a"
}"
`;

exports[`solve Car -- nested 1`] = `
"{
  ab: cons("a", "b"),
  ba: cons("b", "a"),
  a: "a",
  b: "b",
  b2: ?b
}"
`;

exports[`solve Car -- car cons 1`] = `
"{
  a: ?a
}"
`;
